### PR TITLE
feat(models, training): introduce sharding support for spectral losses

### PR DESCRIPTION
**Pull request by ssmmnn11**
support sharding for spectral losses

### CODE REVIEW

**Review by havardhhaugen — DISMISSED** [review]
Really nice work! I did not have time to test the branch explicitly and I'll be away on vacation, but the code changes look good to me. 

Maybe an idea for future work is to allow loss / scaler sharding along arbitrary dimensions to avoid the reshard operation at the end.

File: training/tests/unit/losses/test_loss_function.py [hunk 1]
```diff
@@ -498,6 +498,104 @@ def test_fft2d_spectral_losses_shape_and_validation(target: str) -> None:
...
+    mocker: MockerFixture,
+) -> None:
+    x_dim, y_dim = 4, 4
+    group = object()
+    grid_shard_sizes = [8, 8]
+    channel_shard_sizes = [1, 1]
+    spectral_shard_sizes = [8, 8]
+    local_grid = grid_shard_sizes[0]
+
+    loss = loss_cls(transform="fft2d", x_dim=x_dim, y_dim=y_dim)
```
Comment (sahahner) [inline training/tests/unit/losses/test_loss_function.py]: If we use the SHT transform here, we should also be able to include the AMSE to the tested loss functions without adding another test.

**Review by sahahner — APPROVED** [review]
Looks good to me. Thank you @ssmmnn11 

As far as I see, PR https://github.com/ecmwf/anemoi-core/pull/1189 and https://github.com/ecmwf/anemoi-core/pull/992 will need to be consolidated with a few changes from here now.
